File: NET/JNetCall.Sharp/Server/ServiceHost.cs
// ReSharper disable FunctionNeverReturns
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using JNetCall.Sharp.API;
using JNetProto.Sharp.Beans;
using JNetProto.Sharp.Tools;

namespace JNetCall.Sharp.Server
{
    public sealed class ServiceHost<T> : AbstractHost<T>
    {
        public ServiceHost(Type serviceClass) : base(serviceClass)
        {
        }

        private static ProtoConvert ReadSync(Stream @in, Stream @out, ProtoSettings cfg)
        {
            var convert = new ProtoConvert(@in, @out, cfg);
            const int marker = 0xEE;
            // Send flag
            @out.WriteByte(marker);
            @out.Flush();
            // Receive flag
            while (@in.ReadByte() != marker)
            {
            }
            // Ready!
            return convert;
        }

        public void Open(Stream input, Stream output)
        {
            var inst = CreateInst();
            var methods = inst.GetType().GetMethods();
            using var proto = ReadSync(input, output, _config);
            while (proto.ReadObject<MethodCall>() is var call)
            {
                HandleCall(inst, methods, call, proto);
            }
        }
    }
}

File: JVM/jnetcall-java/src/main/java/jnetcall/java/server/ServiceHost.java
package jnetcall.java.server;

import jnetcall.java.api.MethodCall;
import jnetproto.java.beans.ProtoConvert;
import jnetproto.java.beans.ProtoSettings;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public final class ServiceHost<T> extends AbstractHost<T> implements AutoCloseable {

    public ServiceHost(Class<T> serviceClass) {
        super(serviceClass);
    }

    private static ProtoConvert readSync(InputStream in, OutputStream out, ProtoSettings cfg)
            throws IOException {
        var convert = new ProtoConvert(in, out, cfg);
        final int marker = 0xEE;
        // Send flag
        out.write(marker);
        out.flush();
        // Receive flag
        while (in.read() != marker) ;
        // Ready!
        return convert;
    }

    public void open(InputStream input, OutputStream output) throws Exception {
        var inst = createInst();
        var methods = inst.getClass().getMethods();
        try (var proto = readSync(input, output, config)) {
            MethodCall call;
            while ((call = proto.readObject(MethodCall.class)) != null) {
                handleCall(inst, methods, call, proto);
            }
        }
    }

    @Override
    public void close() throws Exception {
        super.close();
    }
}